### DISCUSSION TITLE
How do I debug a piggy bank add amount failure?

### DISCUSSION THREAD

**Discussion by elliot-gh** [firefly-iii/firefly-iii#7463] [Firefly III] [answered]
So basically, some residual $0.01 interest got added to a default asset account, which I then transferred to a savings account, which I then wanted to add to a larger existing piggy bank that I had. However, when I try add the $0.01 to the piggy bank, I get a red banner that says "Error! Could not add $0.01 to "[my piggy bank name here]". When I check the logs, I don't get anything useful and the only relevant line I see is `[2023-05-05 09:51:36] local.ERROR: Cannot add .01 because canAddAmount returned false.`. The amount should definitely addable, and the piggy bank UI does show $0.01 under "Left for piggy banks".

I then tried bypassing the default asset account and adding that $0.01 deposit to my savings account directly, but I am still seeing the above errors. I did some quick checking on the `piggy_bank_events` and `transactions` tables for those specific transactions, and it seems to me that nothing is wrong there in terms of transaction amounts (amount only has two decimal points as expected). Are there any other tables I should be looking at?

I couldn't reproduce this on the demo instance (adding $0.01 deposit to a default account, transferring that to a savings account, and then adding $0.01 to a piggy bank linked to that savings account). Here is my debug page, but I have been seeing this for a few v6 versions now:

Debug information generated at 2023-05-05 00:00:00 America/Los_Angeles for Firefly III version **6.0.9**.

| Scope | Version |
| --- | --- |
| Firefly III | 6.0.9 |
| Firefly III API | 2.0.1 |
| Build | 466, 29-04-2023 06:55:07 CEST |
| DB version | 19 (exp. 19) |
| PHP | `8.2.4` |
| Host | `Linux` |

| System info | Value |
| --- | --- |
| System TZ | America/Los_Angeles |
| Browser TZ | America/Los_Angeles |
| App environment | local |
| App debug mode | false |
| App cache driver | file |
| App logging | info, stack |
| Display errors | Off |
| Error reporting | ALL errors |
| Max upload | 67108864 (64 MB) |
| Interface | apache2handler |
| Default language | en_US |
| Default locale | equal |
| BCscale | 12 |
| DB drivers | mysql, pgsql, sqlite |
| Current driver | mysql |
| Trusted proxies (.env) | `172.28.0.1` |

| User info | Value |
| --- | --- |
| Login provider | eloquent  |
| User guard | web  |
| Headers | N/A, N/A  |
| Stateful domains |  |
| Session start | 2023-05-01 00:00:00 |
| Session end | 2023-05-31 23:59:59 |
| Session first | 2020-08-31 00:00:00 |
| User ID | 1 |
| User language | en_US |
| User locale | en_US |
| Attempt at "en_US.utf8" | 'en_US.utf8' |
| Attempt at "en_US.UTF-8" | 'en_US.UTF-8' |
| User agent | Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36 |

**Comment by elliot-gh** [comment, marked as answer]
Okay, so I continued to look at potential floating point issues, and then ran this query against the `transactions` table to find any other transactions that don't cleanly have only two digits in the decimals:
```SQL
SELECT *
FROM transactions
WHERE CAST(amount AS CHAR) REGEXP '\\.[\\d]{2}[1-9]+';
```

I found 16 out of 10640 transactions that had floating point issues (for example, amount ending with .159999999999 instead of .16), and two of them were related to the savings account in question. I manually did an update/set to make the amount end with .16 and it seems like the piggy bank can finally add the $0.01. I'll probably go back and manually update the other rows just in case.

**Reply by JC5** [reply]
I've seen this problem pop up with a lot of people (including myself) and I have not found the root cause yet. Firefly III at least, does not use floating point values.
